operator/src/main/java/oracle/kubernetes/operator/logging/OncePerMessageLoggingFilter.java

```diff
@@ -0,0 +1,56 @@
+// Copyright 2019, Oracle Corporation and/or its affiliates.  All rights reserved.
+// Licensed under the Universal Permissive License v 1.0 as shown at
+// http://oss.oracle.com/licenses/upl.
+
+package oracle.kubernetes.operator.logging;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** A LoggingFilter that logs each log message, which are typically message keys, at most once */
+public class OncePerMessageLoggingFilter implements LoggingFilter {
+
+  // allow all messages to be logged when filtering is off
+  volatile boolean filtering = false;
+
+  List messagesLogged = new ArrayList();
+
+  /**
+   * Turn on filtering of log messages and skip logging of messages that have already been logged
+   */
+  public void setFilteringOn() {
+    filtering = true;
+  }
+
+  /**
+   * Turn off filtering of log messages and allow all log messages to be logged
+   *
+   * @param resetLogHistory Whether to also clear the history of messages logged so those messages
+   *     will be logged again next time filtering is turned on
+   */
+  public void setFilteringOff(boolean resetLogHistory) {
+    filtering = false;
+    if (resetLogHistory) {
```

Comment:
Using the same name for both the flag and the method is a bit confusing. The long descriptive comment in the Javadoc is a clue that the parameter isn't as obvious as it could be.

Reply:
Should setFilteringOff(boolean resetLogHistory) be separated into 2 method - setFilteringOff() and resetLogHistory()?

Reply:
That would certainly be easier to understand. If doing both is something that happens a lot, you could either have a third method which does both, or have both methods return *this* and then call them both something like this:

```
filter.setFiltering(false).resetLogHistory();
```